src/ProducerEventProcessor/UploaderClientCaptureEventCollectorTest.cpp

```diff
@@ -0,0 +1,191 @@
+  event.mutable_interned_string()->set_intern(str);
+  return event;
+}
+
+ClientCaptureEvent CreateCaptureFinishedEvent() {
+  ClientCaptureEvent event;
+  event.mutable_capture_finished()->set_status(orbit_grpc_protos::CaptureFinished::kSuccessful);
+  return event;
+}
+
+class UploaderClientCaptureEventCollectorTest : public testing::Test {
+ protected:
+  // Set `early_stop_event_count` to be >= 2 if we want to call `Stop()` early with producing only
+  // `early_stop_event_count` events. Here we require ">= 2" as an early stop will create at
+  // least two capture events: one `InternedStringCaptureEvent` created when stop_called_ becomes
+  // true, and one `CaptureFinishedEvent`.
+  void Start(size_t event_count, uint64_t produce_event_every_ms, uint64_t upload_events_every_ms,
+             size_t early_stop_event_count = 0) {
+    event_count_ = event_count;
+    produce_event_every_ms_ = produce_event_every_ms;
+    upload_events_every_ms_ = upload_events_every_ms;
+    early_stop_event_count_ = early_stop_event_count;
+
+    data_produce_thread_ =
+        std::thread(&UploaderClientCaptureEventCollectorTest::ProduceCaptureData, this);
+    data_upload_thread_ =
+        std::thread(&UploaderClientCaptureEventCollectorTest::UploadCaptureData, this);
+  }
+
+  void Stop() {
+    absl::MutexLock lock{&stop_called_mutex_};
+    stop_called_ = true;
+  }
+
+  void TearDown() override {
+    {
+      absl::MutexLock lock{&stop_called_mutex_};
+      CHECK(stop_called_);
+    }
+
+    CHECK(data_produce_thread_.joinable());
+    data_produce_thread_.join();
+
+    CHECK(data_upload_thread_.joinable());
+    data_upload_thread_.join();
+  }
+
+  UploaderClientCaptureEventCollector collector_;
+  size_t total_uploaded_data_bytes_ = 0;
+
+  absl::Mutex upload_finished_mutex_;
+  bool upload_finished_ ABSL_GUARDED_BY(upload_finished_mutex_) = false;
+  absl::Mutex early_stop_mutex_;
+  bool early_stop_ ABSL_GUARDED_BY(early_stop_mutex_) = false;
+
+ private:
+  // Keep producing capture events until stop is called or `event_count_` events are produced with a
+  // specified produce duration. The last produced event will always be the `CaptureFinishedEvent`.
+  void ProduceCaptureData() {
+    size_t added_event_count = 0;
+    bool need_check_early_stop = (early_stop_event_count_ >= 2);
+    bool stopped = false;
+    while (added_event_count < event_count_ - 1 && !stopped) {
+      if (need_check_early_stop && added_event_count + 2 >= early_stop_event_count_) {
+        {
+          absl::MutexLock lock{&early_stop_mutex_};
+          early_stop_ = true;
+        }
+        need_check_early_stop = false;
+      }
+
+      stop_called_mutex_.LockWhenWithTimeout(absl::Condition(&stop_called_),
+                                             absl::Milliseconds(produce_event_every_ms_));
+      collector_.AddEvent(CreateInternedStringCaptureEvent(kAnswerKey, kAnswerString));
+      stopped = stop_called_;
```

Comment:
You can consider reducing the scope of `stopped`. Something like this:

```
//...
bool stopped = stop_called_;
stop_called_mutex_.Unlock();

++added_event_count;

if (stopped) break;
```

This way it's a bit easier to read and understand.

Reply:
The `stop_called_mutex_` also guarantees we produce event with the given interval.